subsys/logging/log_backend_cavs_hda.c

```diff
@@ -0,0 +1,263 @@
+ * So regardless of the log output we *always* send 128 bytes. The other
+ * side then can interpret it as needed.
+ */
+static __aligned(128) uint8_t hda_log_buf[CONFIG_LOG_BACKEND_CAVS_HDA_SIZE];
+static volatile uint32_t hda_log_buffered;
+static volatile bool initialized, panic_mode;
+static atomic_t hda_log_lock;
+static struct k_timer hda_log_timer;
+static cavs_hda_log_hook_t hook;
+
+static void hda_log_flush(void)
+{
+	/* Flushing is a best effort, if we fail now we can catch it a little
+	 * later
+	 */
+	if (!atomic_cas(&hda_log_lock, 0, 1)) {
+		return;
+	}
+
+	uint32_t nearest128 = hda_log_buffered & ~((128) - 1);
+
+	if (nearest128 > 0) {
+		hda_log_buffered = hda_log_buffered - nearest128;
+
+		dma_reload(hda_log_dev, hda_log_chan, 0, 0, nearest128);
+
+		if (hook != NULL) {
+			hook(nearest128);
+		}
+	}
+
+	atomic_set(&hda_log_lock, 0);
+}
+
+
+static int hda_log_out(uint8_t *data, size_t length, void *ctx)
+{
+	int ret;
+	struct dma_status status;
+
+	while (!atomic_cas(&hda_log_lock, 0, 1)) {
+	}
+
+	/* Defaults when DMA not yet initialized */
+	uint32_t dma_free = sizeof(hda_log_buf);
+	uint32_t write_pos = 0;
+
+	if (initialized) {
+		ret = dma_get_status(hda_log_dev, hda_log_chan, &status);
+		if (ret != 0) {
+			ret = length;
+			goto exit;
+		}
+
+		/* The hardware tells us what space we have available, and where to
+		 * start writing. If the buffer is full we have no space.
+		 */
+		if (status.free <= 0) {
+			ret = length;
+			goto exit;
+		}
+
+		dma_free = status.free;
+		write_pos = status.write_position;
+	}
+
+	/* Account for buffered writes since last dma_reload
+	 *
+	 * No underflow should be possible here, status.free is the apparent
+	 * free space in the buffer from the DMA's read/write positions.
+	 * When dma_reload is called status.free may be reduced by
+	 * the nearest 128 divisible value of hda_log_buffered,
+	 * where hda_log_buffered is then subtracted by the same amount.
+	 * After which status.free should only increase in value.
+	 *
+	 * Assert this trueth though, just in case.
+	 */
+	__ASSERT_NO_MSG(dma_free > hda_log_buffered);
+	uint32_t available = dma_free - hda_log_buffered;
+
+	/* If there isn't enough space for the message there's an overflow */
+	if (available < length) {
+		ret = length;
+		goto exit;
+	}
+
+	/* Copy over the message to the buffer */
+	uint32_t idx = write_pos + hda_log_buffered;
+
+	if (idx > sizeof(hda_log_buf)) {
+		idx -= sizeof(hda_log_buf);
+	}
+
+	size_t copy_len = (idx + length) < sizeof(hda_log_buf) ? length : sizeof(hda_log_buf) - idx;
+
+	memcpy(&hda_log_buf[idx], data, copy_len);
+
+	/* There may be a wrapped copy */
+	size_t wrap_copy_len = length - copy_len;
+
+	if (wrap_copy_len != 0) {
+		memcpy(&hda_log_buf[0], &data[copy_len], wrap_copy_len);
+	}
+
+	ret = length;
+	hda_log_buffered += length;
+
+exit:
+	atomic_set(&hda_log_lock, 0);
+
+	if ((hda_log_buffered > sizeof(hda_log_buf)/2 || panic_mode) && initialized) {
```

Comment:
This is a race: you're testing outside the lock, so two simultaneous threads trying to test it might flush twice, or plausibly miss it entirely if "panic_mode" or "initialized" get mutated.  Best practice is always to put all access to shared data inside a single lock and be consistent about it.   In practice it looks like the flush is just going to take the same lock anyway, so errors would be weird and very rare if they happened.

Also: you want a kconfig to tune that watermark.  Logging is bursty.  It's possible there are some app configurations that need to start flushing almost immediately so that they have more than half the buffer avaialable for a big stack dump, etc....  This is the kind of number that never works for everyone, someone's going to need to hack it at some point.